Fixes loading models, and saving/loading as Tensorflow Lite models

ilmari99 opened the pull request:
Pros:
- Removed dimensionality assertions in all `_check_and_reshape_inputs` methods, since this was preventing saving models with KAN layers as TF Lite files, since they couldn't handle symbolic tensors
- When deserializing a Conv layer, a double rank argument was passed to the constructor by from_config, raising an error
- When serializing Dense, the basis_activation wasn't serialized correctly. Fixed this.
- All tests in 'demo' pass.

Cons:
- Removing the assertions makes the error messages (incase of incorrect dims) less clear, which is a down side that should be fixed

ZPZhou-lab commented:
Thank you for committing and improving this project!🤗

I accept your suggestion about `from_config()` and I apologize for not considering that the `len()` method would raise errors in TensorFlow Lite. I am not very familiar with TensorFlow Lite.

However, I believe it is necessary to help users clarify the input shapes required for each module to use them correctly (especially for beginners and those unfamiliar with TensorFlow). Therefore, I am trying to continue adding the assert check of shapes and change `ndim = len(shape)` to `ndim = tf.size(shape)`. 

I hope this can help you complete the test in TensorFlow Lite. Looking forward to your test results!😄

If the test passes, I will merge it. If the test fails, we will continue to try to solve the problem.

Best!

ilmari99 commented:
Hello,

I agree, that is important! Unfortunately Python comparison doesn't seem to work in the the compute graph.
For example `ndim == 4` raises an error:
`OperatorNotAllowedInGraphError: Using a symbolic `tf.Tensor` as a Python `bool` is not allowed. ---`

I changed Python comparisons and assertions to "tensorflow.debugging.assert*" calls. This should be more robust.
I also changed getting the Tensor's shapes to use their `.get_shape()` method which also works with static shapes. [Difference between static and dynamic shapes](https://stackoverflow.com/questions/37096225/how-to-understand-static-shape-and-dynamic-shape-in-tensorflow).

The error messages are once again descriptive, for example, a missing channel dimension with Conv1DKAN:
```
    File "/home/ilmari/python/tfkan/tfkan/layers/convolution.py", line 110, in call  *
        inputs, orig_shape = self._check_and_reshape_inputs(inputs)
    File "/home/ilmari/python/tfkan/tfkan/layers/convolution.py", line 310, in _check_and_reshape_inputs  *
        tf.debugging.assert_equal(ndim,

    InvalidArgumentError: Expected min_ndim=3, found ndim=2. Full shape received: (None, 10).
    Condition x == y did not hold element-wise:
    x (conv1dkan/assert_equal_1/x:0) = 
    2
    y (conv1dkan/assert_equal_1/y:0) = 
    3


Call arguments received by layer "conv1dkan" (type Conv1DKAN):
  • inputs=tf.Tensor(shape=(None, 10), dtype=float32)
  • args=<class 'inspect._empty'>
  • kwargs={'training': 'None'}
```
If you want to get rid of that "Condition x == y ....." you can can wrap the tf.debugging.assert* calls in a try-except, and catch tf.errors.InvalidArgumentError.

I hope it's good now, let me know if not!

ZPZhou-lab commented:
Thank you for providing the method. In my experiment, these changes were successfully tested and I hope they meet your needs.

I moved the assertion to the parent class `ConvolutionKAN` and called it with a unified interface `self._check_input_tensor_shape()` to make the code more compact.

I will merge this commit. Thank you again for your efforts!🤗

Best😀
